### PR TITLE
Add a startup validation to check for compliance with client ID tenant unification

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
### Proposed changes in this pull request
Capability to create multiple apps with the same client ID in different tenants is provided with https://github.com/wso2/product-is/issues/16370. However this feature requires certain pre-conditions.
- Enable tenant qualified URLs
- Enable tenanted sessions
- Altering DB constraint in `IDN_OAUTH_CONSUMER_APPS` table to make <clientID, tenantID> unique.

These conditions are by default enabled in the pack and feature doesn't support backward compatibility without altering the DB constraint.

This PR is to add a pre-validation to the OAuth component so that server won't startup correctly when the prerequisites doesn't match.

### Related issue
- https://github.com/wso2/product-is/issues/17076

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 1]
```diff
@@ -1720,4 +1724,47 @@ private void handleRequestForANonExistingApp(String appName) throws InvalidOAuth
...
+            throws IdentityOAuth2Exception {
+
+        String sqlQuery = SQLQueries.OAuthAppDAOSQLQueries.CHECK_CONSUMER_KEY_CONSTRAINT_ON_CONSUMER_APPS_TABLE;
+        String tableName = CONSUMER_APPS_TABLE_NAME;
+        String keyColumnName = CONSUMER_KEY;
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            DatabaseMetaData metaData = connection.getMetaData();
+
+            if (metaData.getDatabaseProductName().toLowerCase().contains("db2")) {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java:1744]: There was a util class to check the DB type, please use that

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 2]
```diff
@@ -5036,4 +5036,20 @@ public static boolean isTokenPersistenceEnabled() {
         }
         return OAuth2Constants.DEFAULT_PERSIST_ENABLED;
     }
+
+    /**
+     * Utility method to check if server compatible with client ID tenant unification.
+     *
+     * @return true if compliant.
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java:5043]: without just mentioning compliant, better to have little details when it is compliant and when not

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java (hunk 1 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Fixed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java (hunk 2 above)
Comment (ThaminduDilshan) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java:5046]: Fixed

**Review by pulasthi7 — DISMISSED** [review]

**Comment by ThaminduDilshan** [comment]
TODO: Need to fix the PR builder. It seems the integration tests added for the same feature causes this failure

**Review by sahandilshan — APPROVED** [review]

**Pull request merged**